File: engineering/network-programming/01-tcp-echo/skeleton/go/main.go
package main

import (
	"fmt"
	"time"
)

// runServer 启动 Echo Server，返回 listener 地址。
func runServer() (string, error) {
	// TODO: 监听 TCP，端口设为 0（让 OS 分配）
	// TODO: 在 goroutine 中 accept 循环，每个连接一个 goroutine 处理
	// 处理逻辑：逐行读取，如果不是 "quit" 则回写；是 "quit" 则关闭连接
	return "", fmt.Errorf("not implemented")
}

// runClient 连接 addr，发送 messages，返回 server 的回复列表。
func runClient(addr string, messages []string) ([]string, error) {
	// TODO: 连接 server
	// TODO: 逐行发送 messages，逐行读取回复
	// TODO: 发送 "quit" 并关闭连接
	return nil, fmt.Errorf("not implemented")
}

func main() {
	addr, err := runServer()
	if err != nil {
		fmt.Printf("FAIL: server start: %v\n", err)
		return
	}
	time.Sleep(100 * time.Millisecond) // 给 server 一点时间启动

	messages := []string{"hello", "world", "quit"}
	replies, err := runClient(addr, messages)
	if err != nil {
		fmt.Printf("FAIL: client: %v\n", err)
		return
	}

	// 期望 replies 为 ["hello", "world"]（quit 不回显）
	expected := []string{"hello", "world"}
	if len(replies) != len(expected) {
		fmt.Printf("FAIL: expected %d replies, got %d\n", len(expected), len(replies))
		return
	}
	for i := range expected {
		if replies[i] != expected[i] {
			fmt.Printf("FAIL: reply[%d]=%q, expected %q\n", i, replies[i], expected[i])
			return
		}
	}
	fmt.Println("PASS: tcp echo correct")
}

File: engineering/network-programming/01-tcp-echo/skeleton/python/main.py
# 模型：threading
import socket
import threading
import time

def handle_client(conn, addr):
    # TODO: 逐行读取客户端数据；如果不是 "quit" 则回写；否则关闭连接
    pass

def run_server():
    # TODO: 创建 TCP socket，绑定 0 端口，listen
    # TODO: 在后台线程中 accept 循环
    # TODO: 返回实际监听的地址 (host, port)
    pass

def run_client(addr, messages):
    # TODO: 连接 server，逐行发送 messages，逐行读取回复
    # TODO: 返回回复列表（不包含 quit 的回复）
    pass

def main():
    addr = run_server()
    time.sleep(0.1)

    messages = ["hello", "world", "quit"]
    replies = run_client(addr, messages)

    expected = ["hello", "world"]
    if replies == expected:
        print("PASS: tcp echo correct")
    else:
        print(f"FAIL: expected {expected}, got {replies}")

if __name__ == "__main__":
    main()
